interceptor/middleware/counting.go

```diff
@@ -0,0 +1,82 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/go-logr/logr"
+
+	"github.com/kedacore/http-add-on/pkg/k8s"
+	"github.com/kedacore/http-add-on/pkg/queue"
+	"github.com/kedacore/http-add-on/pkg/util"
+)
+
+type Counting struct {
+	queueCounter    queue.Counter
+	upstreamHandler http.Handler
+}
+
+func NewCountingMiddleware(queueCounter queue.Counter, upstreamHandler http.Handler) *Counting {
+	return &Counting{
+		queueCounter:    queueCounter,
+		upstreamHandler: upstreamHandler,
+	}
+}
+
+var _ http.Handler = (*Counting)(nil)
+
+func (cm *Counting) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r = util.RequestWithLoggerWithName(r, "CountingMiddleware")
+	ctx := r.Context()
+
+	defer cm.countAsync(ctx)()
+
+	cm.upstreamHandler.ServeHTTP(w, r)
+}
+
+func (cm *Counting) countAsync(ctx context.Context) func() {
+	signaler := util.NewSignaler()
+
+	go cm.count(ctx, signaler)
+
+	return func() {
+		go signaler.Signal()
+	}
```

Comment:
Could we have perf problems creating 2 routines for each request? Can't we just avoid the signaler doing something like:
```golang
return func() {
		go cm.count(ctx, sw)
	}
```

Or even if maybe making it sync

Reply:
For this one, I also wouldn't like to perform the operation synchronously. We are locking/unlocking a mutex and operating on a map, this takes time.